src/inmanta/deploy/scheduler.py

```diff
@@ -433,6 +442,50 @@ async def _run_for_agent(self, agent: str) -> None:
 
             self._work.agent_queues.task_done(agent, task)
 
+    async def _start_compliance_check(self, resource: ResourceIdStr) -> None:
+        """
+        This method is expected to be called when the given resource has just finished deploying,
+        no matter the result itself i.e. success or failure.
+        """
+        if (previous_task := self.resource_periodic_deploys.get(resource)) is not None:
+            previous_task.cancel()
+
+        self.resource_periodic_deploys[resource] = self.create_recurring_compliance_check(resource)
+
+    def create_recurring_compliance_check(self, resource: ResourceIdStr) -> asyncio.Task[None]:
+        """
+        Schedule recurring deploys for the given resource every <scheduler_resource_compliance_check_window> seconds
+        and return the associated task. This task will be canceled and scheduled again when the next deploy finishes
+        for this resource.
+        """
+
+        async def delay(coro: Coroutine[Any, Any, None], seconds: float) -> None:
+            """
+            Helper function to run a coroutine after a delay
+            """
+            await asyncio.sleep(seconds)
+            await coro
+
+        async def _trigger_deploy(resource: ResourceIdStr) -> None:
+            """
+            Check that the given resource is in a known good state with no pending updates and schedule a repair if need be
+            """
+            to_deploy: set[ResourceIdStr] = set()
+            to_deploy.add(resource)
+            # TODO Add dependents ?
```

Comment:
This is the big one right! 

For repair, I think we should only be concerned with individual resources. We repair the one-by-one, independently. They respond only to the know states.

For deploy, it becomes more subtle. 
If we define the deploy as the minimal interval between subsequent deploys of a resource, it is clear: anything in the dirty set gets an individual re-try timer. They will naturally line up in the correct order. (except that a pure skip for dependencies is also a deploy for the scheduler, making the external observable behavior complicated again)

If we define it as the retry interval for individual dirty items, we can just run the timer for individual items. they will also naturally line up (deepest ones will expire first and 'block' the others, when they are done, they will immediately queue them, resetting the timer).

If we define it as the retry interval for all resources, we arrive back to the mechanism we already had. 

Retry for a disconnected subtree is what is suggested here, feels a bit weird to me?


I think we need a clear storry here, that we want to present to the wider team

Reply:
Regarding how we define it, I assumed we were talking "he retry interval for individual dirty items" in this iteration.

As to "they will naturally line up", I'm not sure that is the case (in this interpretation / definition). e.g. consider a resource `host` in the common (not the right name but you get the point) resource set and two `file` resources `hello` and `world` that depend on that `host`, each in their own resource set. We first export and release the `hello` resource set, causing `host` and `hello` to be scheduled for deploy. `host` fails, `hello` is skipped. Sometime later we export and release the `world` resource set, causing `host` and `world` to be scheduled. `host` fails again and `world` is skipped. Now the retry times for `host` and `world` are out of sync, unless we just keep stacking triggers for the same resource, but that doesn't seem right to me either.

The worst instance of this issue would occur with a long retry interval, and the two resource deploys only a few seconds after one another. This would cause `hello` to be skipped again, then a few seconds after `host` and `world` might deploy successfully, and `hello` would have to wait for almost the entire retry interval once more.

> Retry for a disconnected subtree is what is suggested here, feels a bit weird to me?

I think all approaches have something to say for them. This one solves some of the issues that others have, but I see your point.

> I think we need a clear storry here, that we want to present to the wider team

I wholeheartedly agree. We probably should have gotten this more clear earlier. We should start from what makes sense conceptually and go from there. Perhaps we could do the simple thing where we have simple per-resource triggers, and then address the issue with the out-of-order retries through a different mechanism, e.g. (draft idea) force "success event propagation" towards skipped resources, even if `receive_events=false`. This would cause skipped resources to be retried when they become unblocked. We'd probably want to move ahead the plan to differentiate between default `deploy` skip and explicit handler skip.

Reply:
Moved discussion to [Slack thread](https://inmanta.slack.com/archives/CKRF0C8R3/p1730735805023259)

Reply:
I'm leaving this one unresolved because I'm pretty sure it's linked in a few places and GitHub does not handle links to resolved discussions well.